packages/flutter/lib/src/material/button_style.dart

```diff
@@ -315,6 +323,22 @@ class ButtonStyle with Diagnosticable {
   /// ```
   final InteractiveInkFeatureFactory? splashFactory;
 
+  /// Creates a widget that becomes the child of the button's [Material]
+  /// and whose child is the rest of the button, including the button's
+  /// `child` parameter.
+  ///
+  /// By default the returned widget is clipped to the Material's [ButtonStyle.shape].
+  final ButtonLayerBuilder? backgroundBuilder;
+
+  /// Creates a Widget that contains the button's child parameter which is used
+  /// instead of the button's child.
+  ///
+  /// The returned widget is clipped by the button's
+  /// [ButtonStyle.shape] inset by the button's [ButtonStyle.padding]
+  /// and aligned by the button's [ButtonStyle.alignment].
+  /// [ButtonStyle.shape].
+  final ButtonLayerBuilder? foregroundBuilder;
```

Comment:
After reading the two docs on foregroundBuilder and backgroundBuilder it really isn't super-clear to me what the difference would be and when to use which one (other then through trail and error). Maybe the docs can be expanded a little bit to work out the differences? There should probably also be a "See also" section for each that links to the other one?

Reply:
The foreground builder inherits the same constraints as the button's child (i.e. they include the button's padding), while the background builder constraints are the same as the button's material - i.e. the button's overall bounds. I will try and make this clearer.